### ISSUE TITLE
"too much recursion" error when creating a Record type from an instance of another Record

### ISSUE THREAD

**Issue by joepie91** [immutable-js/immutable-js#1690]
### What happened

When creating a new Record type, and passing in an instance of *another* Record as the schema / default values, Immutable will end up in an infinite recursion loop, yielding the following error:

![selection_919](https://user-images.githubusercontent.com/1663259/52636394-bd8f4500-2ecc-11e9-9ed7-21ab0bfee0db.png)

I don't know whether it conceptually makes sense to try to do this (see the repro below); in my case the 'nested Record' structure was caused by a mistake during refactoring.

However, the current error is completely unclear about the cause of the issue, and provides no real way to track down the cause. I ended up spending several hours trying to find the root cause.

__Expected behaviour:__ Either produce an understandable error with concrete guidance (if this is not supposed to be supported), or work correctly without breaking (if it *is* supposed to be supported).

### How to reproduce

Repro:

```js
const immutable = require("immutable");

let Foo = immutable.Record({foo: "bar"});
let fooInstance = Foo();
let Bar = immutable.Record(fooInstance);
let barInstance = Bar();
```

Tested in Firefox 64.0 with latest Immutable.js (4.0.0-rc.12).

**Comment by ttamminen** [comment]
I think I know what you tried to achieve. Maybe there should be an error or something, but I think this is more clear.

```
const immutable = require("immutable");

let fooSchema = {foo: "bar"};
let Foo = immutable.Record(fooSchema);
let fooInstance = Foo({ foo: "myFoo" });
let Bar = immutable.Record(fooSchema);
let barInstance = Bar({ foo: 'barFoo' });
```

Or, if you really want the Bar default value to be what fooInstance has.

```
let Foo = immutable.Record({foo: "bar"});
let fooInstance = Foo();
let Bar = immutable.Record(fooInstance.toJS());
let barInstance = Bar();
```

Which in this case would be a value "bar" for `foo` property.

**Comment by joepie91** [comment]
Nah, I didn't actually intend to have two separate Record types and have one contain the other; I just erroneously tried to wrap an object in `immutable.Record(...)` twice.

The problem is that it didn't produce an understandable error, and therefore it took a *long* time to figure out that the problem was an erroneous double call of `immutable.Record`.

**Comment by jdeniau** [comment]
Thank you for your bug report. The immutable-js oss fork will soon be merged and a fix for this issue has been included in #1833. Once this PR is merged, this issue will be resolved in the main branch. We will then do our best to to release the `4.0.0` version.

Commit reference: https://github.com/immutable-js/immutable-js/pull/1833/commits/9c7e5cbf569ae9d987dd99f1b3362aa5184e4ccb
